### PR TITLE
[AArch64] Optimize vector fmul(sitofp/uitofp, 1/2^N) -> scvtf/ucvtf

### LINKED ISSUE #94909: [AArch64] Missed optimization converting vector fmul(sitofp(x), 0.5) -> scvtf(x, 2)
In #91924, we removed the existing combine for vector `fdiv(sitofp(x), 2) -> scvtf(x, 2)` etc. because it would almost never trigger: instcombine converts all fdiv operations by powers of two to fmul.  However, the idea behind the optimization is still useful: for the given conversion operations, we can eliminate the multiply.   We should try to restore this optimization so it works on vector inputs.  (This optimization is already implemented for scalar inputs.)
https://godbolt.org/z/c8qx7v4b4

Marking this good-first-issue because the code that was removed can be used as a starting point for this.

### PR DESCRIPTION
**Pull request by jph-13**
When a vector integer-to-float conversion is followed by a multiply with a
  reciprocal power-of-two constant, we can fold both operations into a single
  SCVTF or UCVTF instruction with a fixed-point shift operand.

  For example, `fmul(sitofp(v2i32 x), <0.5, 0.5>)` becomes `scvtf.2s v0, v0, #1`.

  This is a reworked version with several improvements over the original submission:
  - Rewrite the C++ operand matcher to share implementation with the existing
    `SelectCVTFixedPointVec` (MOVIshift, FMOV, and DUP handling with correct
    truncation for f16)
  - Add `uitofp`/`ucvtf` patterns via a `CVTFRecipPat` multiclass
  - Add full GlobalISel support (`GIComplexOperandMatcher` + renderer)

  Supported vector types: `v2f32`, `v4f32`, `v2f64`, `v4f16`, `v8f16`.

  Fixes #94909

### CODE REVIEW

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp [hunk 1]
```diff
@@ -19250,6 +19253,153 @@ static SDValue performFpToIntCombine(SDNode *N, SelectionDAG &DAG,
...
+  BuildVectorSDNode *BV = cast<BuildVectorSDNode>(ConstVec);
+  int32_t IntrinsicC = getUniformFPSplatLog2(BV, FloatBits + 1);
+
+  // Handle cases where it's not a power of two, or is 2^0.
+  if (IntrinsicC == -1 || IntrinsicC == 0)
+    return SDValue();
+
+  // Check if IntrinsicC is within the valid range [1, FloatBits].
+  // The 's' value must be in [1, FloatBits].
+  if (IntrinsicC <= 0 || IntrinsicC > FloatBits)
```
Comment (stephentyrone) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:19368]: This check makes the preceding check entirely redundant; I think that's OK, if you just want to be super explicit.

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp [hunk 2]
```diff
@@ -19250,6 +19253,153 @@ static SDValue performFpToIntCombine(SDNode *N, SelectionDAG &DAG,
...
+    return SDValue();
+
+  MVT IntTy = Op.getOperand(0).getSimpleValueType().getVectorElementType();
+  int32_t IntBits = IntTy.getSizeInBits();
+  if (IntBits != 16 && IntBits != 32 && IntBits != 64)
+    return SDValue();
+
+  MVT FloatTy = N->getSimpleValueType(0).getVectorElementType();
+  int32_t FloatBits = FloatTy.getSizeInBits();
+  if (FloatBits != 32 && FloatBits != 64)
```
Comment (stephentyrone) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:19352]: IIRC scvtf supports half-precision as well.

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp [hunk 3]
```diff
@@ -19250,6 +19253,153 @@ static SDValue performFpToIntCombine(SDNode *N, SelectionDAG &DAG,
...
+    return SDValue();
+
+  if (IntBits > FloatBits)
+    return SDValue();
```
Comment (stephentyrone) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:19356]: Maybe worth a comment here noting that we cannot just convert to a wider float type and then narrow because it would introduce a potential double-rounding.

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp [hunk 4]
```diff
@@ -19250,6 +19253,153 @@ static SDValue performFpToIntCombine(SDNode *N, SelectionDAG &DAG,
...
+    if (!Val.bitwiseIsEqual(FirstVal))
+      return -1;
+  }
+
+  // Reject zero, NaN, or negative values
+  if (FirstVal.isZero() || FirstVal.isNaN() || FirstVal.isNegative())
+    return -1;
+
+  // Get raw bits
+  APInt Bits = FirstVal.bitcastToAPInt();
```
Comment (stephentyrone) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:19283]: Is there a reason why we do all the rest of this in terms of bit patterns instead of doing `frexp` on the APFloat value to extract the exponent and significand and working with those?

File: llvm/lib/Target/AArch64/AArch64ISelLowering.cpp [hunk 5]
```diff
@@ -19250,6 +19253,153 @@ static SDValue performFpToIntCombine(SDNode *N, SelectionDAG &DAG,
...
+
+  if (static_cast<unsigned>(Log2) > MaxExponent)
+    return -1;
+
+  return Log2;
+}
+
+/// Fold a floating-point multiply by power of two into fixed-point to
+/// floating-point conversion.
+static SDValue performFMulCombine(SDNode *N, SelectionDAG &DAG,
```
Comment (stephentyrone) [inline llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:19322]: Maybe worth a comment that this optimization would not be valid for a constrained FMul intrinsic, because sitofp rounds according to the default rounding mode (RN) but scvtf uses the rounding mode from FPCR.

**Comment by jph-13** [comment]
I will take a look at the tablegen impls. Maybe that will help me understand why I am having issues with half too.

**Comment by jph-13** [comment]
I just resolved all the original flags since the new implementation is very different. I did try to get f16 working but I became very confused. As of now it doesn't appear to have a match in TD. I started creating one but I am not sure if I shold replace all the round tripping or not. So I figured I would see if we could get this in, then maybe try another pass at half later.

**Comment by jph-13** [comment]
This is incomplete, but I could really use some help. I have never touched tablegen before this and feel I am making it way too difficult. 

I am having specific issues with the smaller registers and sizes (are my matchers too restrictive?). I left in commented out vNi16 code for I can't get any of it to not conflict. The v1i32 and v1i64 are also escaping me. As for the f16 implementations, it seems I need to handle a sext, but I figure the rest should be working first. 

I could use any comments or guidance folks have.

Thanks.

File: llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp [hunk 6]
```diff
@@ -3952,6 +3960,156 @@ static bool checkCVTFixedPointOperandWithFBits(SelectionDAG *CurDAG, SDValue N,
...
 }
 
+static bool checkCVTFixedPointOperandWithFBitsForVectors(SelectionDAG *CurDAG,
+                                                         SDValue N,
+                                                         SDValue &FixedPos,
+                                                         unsigned RegWidth,
+                                                         bool isReciprocal) {
+
+  // Fast Path
+  if (N.getOpcode() == ISD::BUILD_VECTOR) {
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp]: I think this can use isConstantSplatVector to check for BUILD_VECTOR or SPLAT, and then some extra logic to check for DUP of a constant or MOVIshift.

File: llvm/lib/Target/AArch64/AArch64InstrFormats.td [hunk 7]
```diff
@@ -5407,6 +5433,102 @@ class BaseIntegerToFPUnscaled<bits<2> rmode, bits<3> opcode,
   let Inst{4-0}   = Rd;
 }
 
+multiclass IntegerToFPVector<
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/AArch64InstrFormats.td]: This looks like it is defining an instruction. We just need a Pat to convert the `mul(sint_to_fp(` to a.. SCVTFv2i32_shift I think. The new pattern will need to be defined after the instruction is, which mean somewhere after the `defm SCVTF: SIMDVectorRShiftToFP<..>`.

File: llvm/lib/Target/AArch64/AArch64InstrFormats.td [hunk 8]
```diff
@@ -5407,6 +5433,102 @@ class BaseIntegerToFPUnscaled<bits<2> rmode, bits<3> opcode,
...
+//   (SCVTFv2f16_v2i64_V FPR128:$Rn, fixedpoint_recip_v2f16_v2i64:$scale)
+// >;
+
+// def : Pat<
+//   (fmul (sint_to_fp (v2i64 FPR128:$Rn)),
+//         fixedpoint_recip_v2f32_v2i64:$scale),
+//   (SCVTFv2f32_v2i64_V FPR128:$Rn, fixedpoint_recip_v2f32_v2i64:$scale)
+// >;
+
+def : Pat<(fmul(sint_to_fp(v2i64 FPR128:$Rn)),
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/AArch64InstrFormats.td]: When you get this far, the tablegen pattern is a pretty gnarly as it needs to convert match the v2f32 $scale, but use that result as a i32 in the instruction. I'm not sure if there is a better way, but it might need a SDNodeXForm to convert the v2f32->i32, even if the complex pattern returns a i32 SDValue.

Something like this might work, if generalized to more types (and cleaned up a bit):
```
class fixedpoint_recip_vec<ValueType FloatVT> : ComplexPattern<FloatVT, 1, "SelectCVTFixedPosRecipOperandVec<32>", []>;
def fixedpoint_recip_v2f32_v2i32_2 : fixedpoint_recip_vec<v2f32>;
def fixedpoint_recip_vec_xform : SDNodeXForm<timm, [{
  return V;
}]>;
def : Pat<(v2f32 (fmul (sint_to_fp(v2i32 V64:$Rn)), fixedpoint_recip_v2f32_v2i32_2:$scale)),
          (v2f32 (SCVTFv2i32_shift (v2i32 V64:$Rn), (fixedpoint_recip_vec_xform fixedpoint_recip_v2f32_v2i32_2:$scale)))>;
```
The xform is a bit weird, as it just allows the type to change without actually doing anything to the value used. It could probably do with an comment explaining it.

File: llvm/test/Transforms/InstCombine/AArch64/scvtf-div-mul-combine.ll [hunk 9]
```diff
@@ -0,0 +1,71 @@
+; RUN: llc -mtriple=aarch64-linux-gnu -aarch64-neon-syntax=apple -mattr=+fullfp16 -o - %s | FileCheck %s
```
Comment (davemgreen) [inline llvm/test/Transforms/InstCombine/AArch64/scvtf-div-mul-combine.ll]: The test can go in llvm/test/CodeGen/AArch64, and the update_llc_test_checks.py script is useful for auto-generating the CHECK lines, so long as they look correct.

**Review by davemgreen — COMMENTED** [review]
It looks like this got quite far. I've left a few comments inline, tablegen can be finicky at times.

File: llvm/lib/Target/AArch64/AArch64InstrFormats.td (hunk 8 above)
Comment (jph-13) [reply llvm/lib/Target/AArch64/AArch64InstrFormats.td]: Thank you so much for the help. I stole your pattern and it is working excellent. However, I am have having issues with f16 due to the promotion (I believe anyway)

I have:

```
let Predicates = [HasNEON, HasFullFP16] in {
def : Pat<(v4f16 (fmul (sint_to_fp(v4i16 V64:$Rn)), fixedpoint_recip_v4f16_v4i16:$scale)),
          (v4f16 (SCVTFv4i16_shift (v4i16 V64:$Rn), (fixedpoint_recip_vec_xform fixedpoint_recip_v4f16_v4i16:$scale)))>;

def : Pat<(v8f16 (fmul (sint_to_fp(v8i16 FPR128:$Rn)), fixedpoint_recip_v8f16_v8i16:$scale)),
          (v8f16 (SCVTFv8i16_shift (v8i16 FPR128:$Rn), (fixedpoint_recip_vec_xform fixedpoint_recip_v8f16_v8i16:$scale)))>;
}
```

Looking at my unmatched test case:
```
  %vcvt.i = sitofp <4 x i16> %in to <4 x half>
  %div.i = fdiv <4 x half> %vcvt.i, <half 0xH4400, half 0xH4400, half 0xH4400, half 0xH4400> 
  ret <4 x half> %div.i
```

I get:
```
// %bb.0:                               // %entry
        sshll.4s        v0, v0, #0
        fmov.4s v1, #0.25000000
        scvtf.4s        v0, v0
        fcvtn   v0.4h, v0.4s
        fcvtl   v0.4s, v0.4h
        fmul.4s v0, v0, v1
        fcvtn   v0.4h, v0.4s
        ret
.Lfunc_end0:
```
If I am using `-mattr=+fullfp16`, why is it still using single? Anymore hints on how to approach?

Comment (davemgreen) [reply llvm/lib/Target/AArch64/AArch64InstrFormats.td]: That does look like it is being promoted. It doesn't happen here: https://godbolt.org/z/xdq14cT67. Any idea what your test might be doing differently?

Comment (jph-13) [reply llvm/lib/Target/AArch64/AArch64InstrFormats.td]: OK, the error was not with the test but rather with the launch command I had for debugging. I repaired that now, sorry for the silly error. 
```
// %bb.0:                               // %entry
        movi.4h v1, #52, lsl #8
        scvtf.4h        v0, v0
        fmul.4h v0, v0, v1
        ret
.Lfunc_end0:
```
Now I think it is close but I still fail to extract the constant for the f16 cases. Using the patterns above it does match. The op comes in as an `AArch64ISD::NVCAST` which then has a 0 operand that is an `AArch64ISD::MOVIshift`. So I seem to be too deep to figure out if it is a splat or not.

Comment (davemgreen) [reply llvm/lib/Target/AArch64/AArch64InstrFormats.td]: The NVCAST just reinterprets the bits from one type to another (always in LE, unlike a BITCAST that can change the bits around). If the element size is the same it should be OK to trivially look through it at the underlying movi.

**Comment by stephentyrone** [comment]
Looks good, thanks for your persistence. Can you add a couple of test cases for examples where we _can't_ do the transform because the power of two is out of range?

File: llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp [hunk 10]
```diff
@@ -3952,6 +3960,129 @@ static bool checkCVTFixedPointOperandWithFBits(SelectionDAG *CurDAG, SDValue N,
...
+    }
+  } else {
+    // ScalarSourceNode is not a recognized constant type.
+    return false;
+  }
+
+  // --- Perform fixed-point reciprocal check and power-of-2 validation on FVal
+  // --- Normalize f16 to f32 if needed for consistent APFloat operations.
+  if (N.getValueType().getVectorElementType() == MVT::f16) {
+    bool ignored;
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp]: ignored -> Ignored
Can you give move details on what "consistent APFloat operations" mean for f16?

File: llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp [hunk 11]
```diff
@@ -3952,6 +3960,129 @@ static bool checkCVTFixedPointOperandWithFBits(SelectionDAG *CurDAG, SDValue N,
...
+    // This implies that the DAG structure was (DUP (MOVIshift C)) or
+    // (BUILD_VECTOR (MOVIshift C)).
+    isSplatConfirmed = true;
+  } else {
+    return false;
+  }
+
+  // If we reached here, isSplatConfirmed should be true and ScalarSourceNode
+  // should be set. But just in case ...
+  if (!isSplatConfirmed)
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp]: Can remove isSplatConfirmed if the above if/else returns false already?

File: llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp [hunk 12]
```diff
@@ -3952,6 +3960,129 @@ static bool checkCVTFixedPointOperandWithFBits(SelectionDAG *CurDAG, SDValue N,
...
+static bool checkCVTFixedPointOperandWithFBitsForVectors(SelectionDAG *CurDAG,
+                                                         SDValue N,
+                                                         SDValue &FixedPos,
+                                                         unsigned FloatWidth,
+                                                         bool isReciprocal) {
+
+  // N must be a bitcast/nvcast of a vector float type.
+  if (!((N.getOpcode() == ISD::BITCAST ||
+         N.getOpcode() == AArch64ISD::NVCAST) &&
+        N.getValueType().isVector() && N.getValueType().isFloatingPoint())) {
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp]: Could this check that the input type is a vector fp type?
```
if (N.getOpcode() == ISD::BITCAST || N.getOpcode() == AArch64ISD::NVCAST)
  N = N.getOperand(0);
EVT VT = N.getValueType();
if (!N.getValueType().isVector() || !N.getValueType().isFloatingPoint())
  return false;
```
(Or something like it).

File: llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp [hunk 13]
```diff
@@ -3952,6 +3960,129 @@ static bool checkCVTFixedPointOperandWithFBits(SelectionDAG *CurDAG, SDValue N,
...
+    return false;
+  }
+
+  if (N.getNumOperands() == 0)
+    return false;
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp]: I think this is already checked by the `(N.getOpcode() == ISD::BITCAST || N.getOpcode() == AArch64ISD::NVCAST)` above?

File: llvm/test/CodeGen/AArch64/scvtf-div-mul-combine.ll [hunk 14]
```diff
@@ -0,0 +1,173 @@
...
+entry:
+  %vcvt.i = sitofp i64 %in to double
+  %mul.i = fmul double %vcvt.i, 0.02
+  ret double %mul.i
+}
+
+; Vector v2f32 with non-power-of-2 constant
+define <2 x float> @negative_test_v2f32_mul_non_pow2_const(<2 x i32> %in) {
+; CHECK-LABEL: negative_test_v2f32_mul_non_pow2_const:
+; CHECK: fmul.2s
```
Comment (davemgreen) [inline llvm/test/CodeGen/AArch64/scvtf-div-mul-combine.ll]: Can you run update_llc_test_checks.py on the test?

**Review by davemgreen — COMMENTED** [review]
Sorry I missed the update. Nice job getting this working, it looks good. I just have a couple of nitpicks, it looks good to me otherwise.

File: llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp (hunk 11 above)
Comment (jph-13) [reply llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp]: I left this and removed the returns from above. I thought the code is more descriptive that way.

File: llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp (hunk 12 above)
Comment (jph-13) [reply llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp]: But the input is a vector of ints at this point no?

File: llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp (hunk 10 above)
Comment (jph-13) [reply llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp]: This code was a relic from an AI conversation. I don't think the code was needed. Originally I was doing more intermediate  math. I removed the logic.

File: llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp (hunk 12 above)
Comment (jph-13) [reply llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp]: updated as:
```
  SDValue ImmediateNode;
  // N must be a bitcast or nvcast
  if (N.getOpcode() == ISD::BITCAST || N.getOpcode() == AArch64ISD::NVCAST) {
    ImmediateNode = N.getOperand(0);
  } else {
    return false;
  }

  EVT NodeVT = N.getValueType();
  EVT InputVT = ImmediateNode.getValueType();
  // The input must me a vector of int but N must be a floating point
  if (!InputVT.isVector() || !InputVT.isInteger() ||
      !NodeVT.isFloatingPoint())
    return false;
    ```

File: llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp [hunk 15]
```diff
@@ -3952,6 +3960,119 @@ static bool checkCVTFixedPointOperandWithFBits(SelectionDAG *CurDAG, SDValue N,
...
+    APInt SplatUndef;
+    unsigned SplatBitSize;
+    bool HasAnyUndefs;
+    if (BVN->isConstantSplat(SplatValue, SplatUndef, SplatBitSize,
+                             HasAnyUndefs)) {
+      IsSplatConfirmed = true;
+    }
+  } else if (ImmediateNode.getOpcode() == AArch64ISD::MOVIshift) {
+    // This implies that the DAG structure was (DUP (MOVIshift C)) or
+    // (BUILD_VECTOR (MOVIshift C)).
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp]: I think this comment is old and can be removed?

File: llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp [hunk 16]
```diff
@@ -3952,6 +3960,119 @@ static bool checkCVTFixedPointOperandWithFBits(SelectionDAG *CurDAG, SDValue N,
...
+    // floating-point.
+    EVT FloatEltVT = N.getValueType().getVectorElementType();
+
+    if (FloatEltVT == MVT::f32) {
+      FVal = APFloat(APFloat::IEEEsingle(), CI->getAPIntValue());
+    } else if (FloatEltVT == MVT::f64) {
+      FVal = APFloat(APFloat::IEEEdouble(), CI->getAPIntValue());
+    } else if (FloatEltVT == MVT::f16) {
+      auto *ShiftAmountConst =
+          dyn_cast<ConstantSDNode>(ImmediateNode.getOperand(1));
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp]: A DUP would not have a operand(1). For MOVIshift it can use something like this, but it is needed for all types not just f16:
```
    APInt Imm = APInt(NodeVT.getScalarSizeInBits(),
                      ImmediateNode.getConstantOperandVal(0)
                          << ImmediateNode.getConstantOperandVal(1));
```

It might be worth making sure more constants are tested. 2.0 uses FMOV which would need AArch64ISD::FMOV and decodeAdvSIMDModImmType11 I believe. Most of the others like 4.0 seem to use MOVIshift.

**Review by davemgreen — COMMENTED** [review]
Thanks - I think the handling of MOVIshift might need to be tightened up a bit, otherwise it could ignore important certain bits of the constant. Detecting the constant correctly is the most awkward part of this with all the different combos, but luckily it looks like many of them use a similar node type.

File: llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp (hunk 16 above)
Comment (jph-13) [reply llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp]: Oh wow, you are completely correct! My impl is missing much more than it catches. I will expand the tests.

File: llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp [hunk 17]
```diff
@@ -3952,6 +3960,132 @@ static bool checkCVTFixedPointOperandWithFBits(SelectionDAG *CurDAG, SDValue N,
...
+    } else if (FloatEltVT == MVT::f64) {
+      FVal = APFloat(APFloat::IEEEdouble(), Imm);
+    } else if (FloatEltVT == MVT::f16) {
+      FVal = APFloat(APFloat::IEEEhalf(), Imm);
+    } else {
+      // Unsupported floating-point element type.
+      return false;
+    }
+  } else if (auto *CFP =
+                 dyn_cast<ConstantFPSDNode>(ImmediateNode.getOperand(0))) {
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp]: Does this case come up?

File: llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp [hunk 18]
```diff
@@ -3952,6 +3960,132 @@ static bool checkCVTFixedPointOperandWithFBits(SelectionDAG *CurDAG, SDValue N,
...
+                                                         SDValue N,
+                                                         SDValue &FixedPos,
+                                                         unsigned FloatWidth,
+                                                         bool IsReciprocal) {
+
+  SDValue ImmediateNode;
+  // N must be a bitcast, nvcast, or fmov
+  if (N.getOpcode() == ISD::BITCAST || N.getOpcode() == AArch64ISD::NVCAST ||
+      N.getOpcode() == AArch64ISD::FMOV) {
+    ImmediateNode = N.getOperand(0);
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp]: I would suggest that this be a little tighter on what it considers, to prevent any (rare) cases from treating the node with an incorrect type (like looking at operand 0 where it doesn't exist and how FMOV is currently handled.

I would probably make it something like
```
SDValue ImmediateNode = N;
if (N.getOpcode() == ISD::BITCAST || N.getOpcode() == AArch64ISD::NVCAST)
  ImmediateNode = N.getOperand(0);
```

Then check the node type is still a vector (in case it is bitcast to a scalar) `ImmediateNode.getValueType().isVector()`. 
The if below would check `ImmediateNode.getOpcode() == AArch64ISD::FMOV`, not for ISD::Constant, and check its operand (which we know should always be a ConstantSDNode.

That should hopefully make it more resilient to odd cases that don't come up very often, and hopefully a little more clear which nodes equate to what.

**Review by davemgreen — COMMENTED** [review]
Thanks for adding all the tests, that's fantastic. I have a suggestion to make the code slightly more structured in case something happens with weird node types, but it LGTM otherwise.

File: llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp (hunk 17 above)
Comment (jph-13) [reply llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp]: Not in the test cases I have. Another cautionary reiic from an AI conversation. I will drop it.

File: llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp [hunk 19]
```diff
@@ -3952,6 +3960,125 @@ static bool checkCVTFixedPointOperandWithFBits(SelectionDAG *CurDAG, SDValue N,
...
+  SDValue ImmediateNode = N.getOperand(0);
+  if (N.getOpcode() == ISD::BITCAST || N.getOpcode() == AArch64ISD::NVCAST) {
+    // This could have been a bitcast to a scalar
+    if (!ImmediateNode.getValueType().isVector())
+      return false;
+  }
+
+  if (ImmediateNode.getOpcode() == AArch64ISD::DUP ||
+      ImmediateNode.getOpcode() == AArch64ISD::MOVIshift ||
+      ImmediateNode.getOpcode() == AArch64ISD::FMOV ||
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp]: I don't think this can be FMOV now.

File: llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp [hunk 20]
```diff
@@ -3952,6 +3960,125 @@ static bool checkCVTFixedPointOperandWithFBits(SelectionDAG *CurDAG, SDValue N,
...
+
+  APInt Imm;
+  bool IsIntConstant = false;
+  if (ImmediateNode.getOpcode() == AArch64ISD::MOVIshift) {
+    EVT NodeVT = N.getValueType();
+    Imm = APInt(NodeVT.getScalarSizeInBits(),
+                ImmediateNode.getConstantOperandVal(0)
+                    << ImmediateNode.getConstantOperandVal(1));
+    IsIntConstant = true;
+  } else if (ImmediateNode.getOpcode() == ISD::Constant) {
```
Comment (davemgreen) [inline llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp]: Could we end up with:
 - N is some random node with 1 constant operand. Say a MOVImsl.
 - Nothing checks N's opcode, only that is has >= 1 operand.
 - The ImmediateNode turns out to be a ISD::Constant, which is treated as a FMOV operand even though it wasn't.

(Granted a MOVImsl will no naturally be a fp type but it could be some other random operand type to the fmul).

Comment (jph-13) [reply llvm/lib/Target/AArch64/AArch64ISelDAGToDAG.cpp]: Fair point. I fixed my shielding code and removed the FMUL from above.

```
  if (!(ImmediateNode.getOpcode() == AArch64ISD::DUP ||
        ImmediateNode.getOpcode() == AArch64ISD::MOVIshift ||
        ImmediateNode.getOpcode() == ISD::BUILD_VECTOR ||
        ImmediateNode.getOpcode() == ISD::Constant ||
        ImmediateNode.getOpcode() == ISD::SPLAT_VECTOR)) {
    return false;
  }

  if (ImmediateNode.getOpcode() != ISD::Constant) {
    auto *C = dyn_cast<ConstantSDNode>(ImmediateNode.getOperand(0));
    if (!C)
      return false;
  }
```
  
The ISD::Constant logic is needed for ```test_v4f32_div_const_4```

**Comment by davemgreen** [comment]
Hi - sorry for the delay. My plan was to wait until after the release and then get this committed and see how it does. I'm still not 100% on whether it might accidentally handle a node incorrectly if it has a constant operand.

I was running some extra tests though and ran into this case, where this particular constant gets lowered to an movi I think, but the operand is an i32 (due to type legalization) and the fp type is a fp16. It just needs to use `FVal = APFloat(APFloat::IEEEhalf(), Imm.trunc(16));` for fp16 values.
```
define <4 x half> @test_v4f16_div_const_0xH1c04(<4 x i16> %in) {
; CHECK-LABEL: test_v4f16_div_const_0xH1c04:
; CHECK:       // %bb.0: // %entry
; CHECK-NEXT:    mov w8, #7172 // =0x1c04
; CHECK-NEXT:    scvtf.4h v0, v0
; CHECK-NEXT:    dup.4h v1, w8
; CHECK-NEXT:    fmul.4h v0, v0, v1
; CHECK-NEXT:    ret
entry:
  %vcvt.i = sitofp <4 x i16> %in to <4 x half>
  %div.i = fmul <4 x half> %vcvt.i, <half 0xH1c04, half 0xH1c04, half 0xH1c04, half 0xH1c04>
  ret <4 x half> %div.i
}
```

I would also recommend changing the start of the function to something like the code below, just to make it super clear how we derive the immediate for each opcode and it is obvious which opcode leads to what.
```
static bool checkCVTFixedPointOperandWithFBitsForVectors(SelectionDAG *CurDAG,
                                                         SDValue N,
                                                         SDValue &FixedPos,
                                                         unsigned FloatWidth,
                                                         bool IsReciprocal) {
  SDValue ImmediateNode = N;
  if (N.getOpcode() == ISD::BITCAST || N.getOpcode() == AArch64ISD::NVCAST) {
    ImmediateNode = N.getOperand(0);
    // This could have been a bitcast to a scalar
    if (!ImmediateNode.getValueType().isVector())
      return false;
  }

  APInt Imm;
  if (ImmediateNode.getOpcode() == ISD::BUILD_VECTOR) {
    // For BUILD_VECTOR, we must explicitly check if it's a constant splat.
    BuildVectorSDNode *BVN = cast<BuildVectorSDNode>(ImmediateNode.getNode());
    APInt SplatUndef;
    unsigned SplatBitSize;
    bool HasAnyUndefs;
    if (!BVN->isConstantSplat(Imm, SplatUndef, SplatBitSize, HasAnyUndefs) ||
        SplatBitSize != N.getValueType().getScalarSizeInBits())
      return false;
  } else if (ImmediateNode.getOpcode() == AArch64ISD::MOVIshift) {
    EVT NodeVT = N.getValueType();
    Imm = APInt(NodeVT.getScalarSizeInBits(),
                ImmediateNode.getConstantOperandVal(0)
                    << ImmediateNode.getConstantOperandVal(1));
  } else if (ImmediateNode.getOpcode() == AArch64ISD::FMOV) {
    uint8_t EncodedU8 = ImmediateNode.getConstantOperandVal(0);
    uint64_t DecodedBits = AArch64_AM::decodeAdvSIMDModImmType11(EncodedU8);

    unsigned BitWidth = N.getValueType().getVectorElementType().getSizeInBits();
    uint64_t Mask = (BitWidth == 64) ? ~0ULL : ((1ULL << BitWidth) - 1);
    uint64_t MaskedBits = DecodedBits & Mask;

    Imm = APInt(BitWidth, MaskedBits);
  } else if (ImmediateNode.getOpcode() != AArch64ISD::DUP ||
             ImmediateNode.getOpcode() != ISD::SPLAT_VECTOR) {
    auto *CI = dyn_cast<ConstantSDNode>(ImmediateNode.getOperand(0));
    if (!CI)
      return false;
    Imm = CI->getAPIntValue();
  } else {
    return false;
  }
```

**Comment by jph-13** [comment]
Sorry, this had fell in the priority list. I have some extra cycles now, hoping to get accross the line.

**Review by davemgreen — APPROVED** [review]
LGTM - thanks for doing this, it looks great.

Do you have commit access or are you happy for me to merge?

**Comment by jph-13** [comment]
@davemgreen, I do not have commit. Could you please do it? Thank you in advance.
